### PR TITLE
fix(modal): fix bindToController bug

### PR DESCRIPTION
**Pull request by kuitos**
when set bindToController true,props not be bound to controller instance correctly before it initialize,this pointer not be init with $scope props

what I added in modal unit test
```js
var $scope = $rootScope.$new(true);
$scope.foo = 'bar';
open({
  template: '<div>{{test.fromCtrl}} {{test.closeDismissPresent()}} {{test.foo}}</div>',
  controller: function($uibModalInstance) {
  + expect(this.foo).toEqual($scope.foo);
    this.fromCtrl = 'Content from ctrl';
    this.closeDismissPresent = function() {
```
it will make the test fail
```shell
Expected undefined to equal 'bar'.
```
because the props not bound to the this instance of controller correctly before the controller constructor called.

[Review on Reviewable](https://reviewable.io/reviews/angular-ui/bootstrap/5569)

### CODE REVIEW

File: src/modal/modal.js [hunk 1]
```diff
@@ -681,18 +681,28 @@ angular.module('ui.bootstrap.modal', ['ui.bootstrap.stackedMap'])
...
                       ctrlInstance.$dismiss = modalScope.$dismiss;
                       angular.extend(ctrlInstance, providedScope);
-                      if (angular.isFunction(ctrlInstance.$onInit)) {
-                        ctrlInstance.$onInit();
-                      }
+                    }
+
+                    ctrlInstance = ctrlInstantiate();
+
+                    if (modalOptions.bindToController && angular.isFunction(ctrlInstance.$onInit)) {
```
Comment (wesleycho) [inline src/modal/modal.js]: Why is `bindToController` being required here for $onInit usage?

File: src/modal/modal.js [hunk 2]
```diff
@@ -681,18 +681,28 @@ angular.module('ui.bootstrap.modal', ['ui.bootstrap.stackedMap'])
                     ctrlLocals[key] = value;
                   });
 
-                  ctrlInstance = $controller(modalOptions.controller, ctrlLocals);
+                  // the third param will make the controller instantiate later,private api
+                  // @see https://github.com/angular/angular.js/blob/master/src/ng/controller.js#L126
+                  ctrlInstantiate = $controller(modalOptions.controller, ctrlLocals, true);
                   if (modalOptions.controllerAs) {
+
```
Comment (wesleycho) [inline src/modal/modal.js]: Remove extra space

File: src/modal/modal.js (hunk 1 above)
Comment (icfantv) [inline src/modal/modal.js]: Per https://docs.angularjs.org/guide/component, in the Intercomponent Communication section, `$onInit` does NOT require `bindToController` being present.

File: src/modal/modal.js (hunk 2 above)
Comment (kuitos) [inline src/modal/modal.js]: OK

File: src/modal/modal.js (hunk 1 above)
Comment (kuitos) [inline src/modal/modal.js]: yes u are right,I had checked the source code of angular.js,it does not require `controllerAs` too.

**Pull request closed without merging**